vApp properties support

goshkis opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
The change adds support for vApp properties manipulation.
Defined properties are available to guest via OVF environment.
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
lib/ansible/modules/cloud/vmware/vmware_guest.py
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.0 (devel f96d1ee759) last updated 2017/11/05 12:36:33 (GMT -700)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /root/ansible/lib/ansible
  executable location = /root/ansible/bin/ansible
  python version = 2.7.5 (default, Aug  4 2017, 00:39:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
```

Akasurde commented:
@goshkis Could you please add integration test for this change ? you can find starting point [here](https://github.com/ansible/ansible/blob/devel/test/integration/targets/vmware_guest/tasks/create_d1_c1_f0.yml)

goshkis commented:
@Akasurde @dagwieers @dav1x @jctanner @nerzhul @pdellaert
Could you please review this request? Thanks.

goshkis commented:
ready_for_review

Akasurde reviewed (approved):
LGTM

goshkis committed:
Adding support for vApp properties.

    List of supported fields can be found here:
    https://github.com/vmware/pyvmomi/blob/master/docs/vim/vApp/PropertySpec.rst
    https://github.com/vmware/pyvmomi/blob/master/docs/vim/vApp/PropertyInfo.rst

    Example:

                "vapp_properties": [
                    {
                        "id": "some_property_to_remove",
                        "operation": "remove"
                    },
                    {
                        "category": "Backup",
                        "label": "Backup server IP",
                        "id": "backupserver.address",
                        "type": "string",
                        "value": "10.0.0.10"
                    }

                 ]

Akasurde commented:
@pdellaert @tchernomax LGTM, can you please take a look ?

tchernomax commented on test/integration/targets/vmware_guest/tasks/vapp_d1_c1_f0.yml:
```diff
@@ -0,0 +1,113 @@
+      type: string
+      value: prop_value2
+    state: present
+  register: vapp_vm
```
Can you re-define the same property and check the module return "change == false".
Currently it will not… but it should.

goshkis replied:
It seems like vcsim does not support vApp properties: every time I read vm_obj.config.vAppConfig from vcsim vm its empty. Is that possible?

goshkis replied:
Proposed test with re-defining the same property executed on real vCenter Server and passed

tchernomax commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -975,6 +1001,68 @@ def configure_network(self, vm_obj):
+        newPropIndex = max(allKeys) + 1 if allKeys else 0
+
+        for k, v in propsToChange.items():
+            tmp = vim.VAppPropertySpec()
```
Is it not vim.vApp.PropertySpec ?
https://github.com/vmware/pyvmomi/blob/575ab56eb56f32f53c98f40b9b496c6219c161da/docs/vim/vApp/PropertySpec.rst

Akasurde replied:
Nice catch.

goshkis replied:
Both variants are valid, but will change for consistency:

In [330]: vim.VAppPropertySpec()
Out[330]:
(vim.vApp.PropertySpec) {
   dynamicType = <unset>,
   dynamicProperty = (vmodl.DynamicProperty) [],
   operation = <unset>,
   removeKey = <unset>,
   info = <unset>
}

In [331]: vim.vApp.PropertySpec()
Out[331]:
(vim.vApp.PropertySpec) {
   dynamicType = <unset>,
   dynamicProperty = (vmodl.DynamicProperty) [],
   operation = <unset>,
   removeKey = <unset>,
   info = <unset>
}

tchernomax commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -975,6 +1001,68 @@ def configure_network(self, vm_obj):
+                    tmp.info = vappProperties[k]
+                    try:
+                        for infoPropName, infoPropVal in v.items():
+                            setattr(tmp.info, infoPropName, infoPropVal)
```
Can you modify this method so the property get updated only if needed (if changed have to be applied) ?

pdellaert replied:
agreed

goshkis replied:
Fixed

tchernomax commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -975,6 +1001,68 @@ def configure_network(self, vm_obj):
                 self.configspec.deviceChange.append(nic)
                 self.change_detected = True
 
+    def configure_vapp_properties(self, vm_obj):
```
Can you rename some variables ; for example  vappProperties could be vapp_properties_current, propsToChange could be vapp_properties_to_change…
Avoid variable like x, k, v, tmp… unless you are in one line expression like `dict((x.id, x) for i in …)`.

pdellaert replied:
agreed

goshkis replied:
Fixed

tchernomax reviewed (changes requested):
ping @goshkis

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -975,6 +1001,68 @@ def configure_network(self, vm_obj):
+                tmp.operation = 'add'
+                tmp.info = vim.VAppPropertyInfo()
+                tmp.info.key = newPropIndex
+                newPropIndex += 1  # maybe use cycle from itertools?
```
Remove comment

goshkis replied:
Fixed

goshkis committed:
Adressing reviewers comments:
* vm specification updated only if changes have to be applied. I.e. subsequent updates with the same data will not trigger changed state
* Auxilary variables renamed, hope this makes the code more readable
* Integration tests changed - re-adding the same properties test not implemented, but tested on real vcenter deployment

Changes tested on vCenter 6.5 deployment

goshkis committed:
* fixing documentation "version_added" for the feature

goshkis commented:
@pdellaert @tchernomax @Akasurde could you please take another look? Thanks.

tchernomax commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -975,6 +1001,91 @@ def configure_network(self, vm_obj):
+
+            new_vapp_property_spec = vim.vApp.PropertySpec()
+
+            if property_id in vapp_properties_current.keys():
```
no need for `.keys()`.
`if property_id in vapp_properties_current:` is the same as `if property_id in vapp_properties_current.keys():`

goshkis replied:
Fixed

tchernomax commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -975,6 +1001,91 @@ def configure_network(self, vm_obj):
+        new_property_index = max(all_keys) + 1 if all_keys else 0
+
+        for property_id, property_spec in vapp_properties_to_change.items():
+
```
is_vapp_changed = False

goshkis replied:
The flag should be set to True if _any_ property changed, so It can't be inside this cycle.  Otherwise. if no changes to the last property in the list, the vm update will not be triggered. Will move it to 1026.

tchernomax replied:
I think you didn't understand what I was suggesting.

What you do: Add each of the property the user have asked for in new_vmconfig_spec.property, and if at least one of them modify the properties already attached to the VM → apply new_vmconfig_spec
What I propose: Only add to new_vmconfig_spec.property the property the user have asked for that actually modify (add/remove/edit) the properties already attached to the VM. And then, if there is some properties in new_vmconfig_spec.property → apply new_vmconfig_spec.

tchernomax replied:
I think you didn't understand what I was suggesting.

What you do
: Add each of the property the user have asked for in new_vmconfig_spec.property, and if at least one of them modify the properties already attached to the VM → apply new_vmconfig_spec

What I propose
: Only add to new_vmconfig_spec.property the property the user have asked for that actually modify (add/remove/edit) the properties already attached to the VM. And then, if there is some properties in new_vmconfig_spec.property → apply new_vmconfig_spec.

goshkis replied:
Thanks, that makes sense now. Now collecting only properties that needs update, and if the list is not empty, assigning vApp configuration.

tchernomax commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -975,6 +1001,91 @@ def configure_network(self, vm_obj):
+            if not x.get('id'):
+                self.module.fail_json(msg="id is required to set vApp property")
+
+        is_vapp_changed = False
```
move this line 1028

tchernomax commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -975,6 +1001,91 @@ def configure_network(self, vm_obj):
+
+        if is_vapp_changed:
+            self.configspec.vAppConfig = new_vmconfig_spec
+            self.change_detected = True
```
```
…
+
+                is_vapp_changed = True
+
+            new_vmconfig_spec.property.append(new_vapp_property_spec)
+
+        if is_vapp_changed:
+            self.configspec.vAppConfig = new_vmconfig_spec
+            self.change_detected = True
```
→
```
…
+
+                is_vapp_changed = True
+            if is_vapp_changed:
+                new_vmconfig_spec.property.append(new_vapp_property_spec)
+        if new_vmconfig_spec.property:
+            self.configspec.vAppConfig = new_vmconfig_spec
+            self.change_detected = True
```

goshkis replied:
Fixed

tchernomax commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -216,6 +216,16 @@
+    - ' - C(id) (string): Property id - required'
+    - ' - C(value) (string): Property value'
+    - ' - C(type) (string): Value type, string type by default.'
+    - ' - C(operation) (string): add, edit, remove. Required only when removing properties. If property present the parameter will be defaulted to "edit".'
```
From what I understand:
* "add" can only be used for adding a non existing property (it will fail on the second run)
* "edit" can be used to add a non existing property or edit an existing property
* "remove" can be used to remove an existing property and will not fail if the property doesn't exist.

Am I right ? If that so this should explicitly explain in the documentation.

goshkis replied:
@tchernomax, thanks for pointing to this. I agree this is probably misleading or confusing. 
These are the three acceptable values for vim.option.ArrayUpdateSpec.Operation, but for the proposed code If operation is set to anything other than "remove" it's silently ignored.

Actual operation value calculated, thus no failures expected when user specify add for an existent property of trying to edit nonexistent one. For such cases, instead of failing we're just notifying user if changes were applied or not.

So, how about:  
    - ' - C(operation): C(remove). This attribute required only when removing properties.

Thanks.

tchernomax replied:
From your code:
* if the user set operation = add:
  * the prop. is present → FAIL, you don't redefine the operation before sending it to vim.option.ArrayUpdateSpec.Operation
  * the prop. isn't present → OK (it is created)
* if the user set operation = edit:
  * the prop. is present → OK (it is edited)
  * the prop. isn't present → OK (the operation re-defined as 'add')
* if the user set operation = remove:
  * the prop. is present → OK (it is removed)
  * the prop. isn't present → OK (nothing is done)

So there is still one FAIL.

Also I am OK with your proposition to change the operation:
> - ' - C(operation): C(remove). This attribute required only when removing properties.'

goshkis replied:
You definitely right here, thanks. Made changes to both: the failure case, and the documentation.

tchernomax reviewed (changes requested):
ping @goshkis

tchernomax reviewed (changes requested):
ping @goshkis

goshkis committed:
Adressing reviewers comments #2:
* documentation updated with the only meaningful value for "option" attribute - "remove"
* Fixed improperly handled case when user requested "add" operation for existent property
* vApp configuration is updated only with properties that contains changes, not with all properties requested by user

goshkis commented:
ping @tchernomax 
Thanks for detailed review!

tchernomax reviewed (approved):
LGTM

Akasurde commented:
@goshkis Thanks for your contribution.

@tchernomax @pdellaert Thanks for your valuable reviews.
